Implement Redux Thunk

smilee opened the pull request:
## 과제 목표

지금까지는 클라이언트에서 레스토랑을 등록하고 조회했습니다. 이제 실제로 서버에서 레스토랑 목록을 불러와서 보여줘야 합니다! 서버에 요청하는 로직은 비동기 로직으로 기존에 사용하던 액션만으로는 부족합니다. Redux Thunk를 이용해서 서버에서 정보를 불러와서 보여주는 앱을 작성 해주세요.

- TDD로 기능 구현
- 레스토랑 지역 목록 보기 기능
- 레스토랑 카테고리 목록 보기 기능
- 레스토랑 목록 보기 기능
- 커버리지 100% 달성

## API

- 레스토랑 지역 목록 보기: https://eatgo-customer-api.ahastudio.com/regions
- 레스토랑 카테고리 목록 보기: https://eatgo-customer-api.ahastudio.com/categories
- 레스토랑 목록 보기: https://eatgo-customer-api.ahastudio.com/restaurants?region=${regionName}&category=${categoryId}

## 제한 조건

- 모든 기능은 TDD로 구현합니다. 컴포넌트와 reducer에 대한 테스트가 존재해야 합니다. 먼저 테스트를 추가하고 commit 한 뒤 실제 기능을 추가하고 commit합니다. commit 메시지 로그는 아래와 같이 남아야 합니다.

```
Restaurant에 대한 테스트를 추가한다

Restaurant을 구현한다

Restaurant를 리팩토링한다

...
```

- if는 사용할 수 있어도 else 사용하실 수 없습니다. GuardClauses 방법을 사용해주세요.
- switch는 사용하실 수 없습니다.
- let은 사용하실 수 없습니다. const만을 사용하여 문제를 해결해주세요.
- 함수 이름과 변수 이름에 줄임말은 사용하실 수 없습니다. 길더라도 명확하게 써주세요.
- indent(인덴트, 들여쓰기) depth를 1로 유지해주세요. 예를 들어, for문 안에 if문이 있으면 indent depth는 2입니다. depth를 줄이는 좋은 방법은 함수(또는 메소드)를 분리하면 됩니다.

## 요구사항

- [x] ESLint를 돌린 결과 아무런 문제가 없어야 합니다.
- [x] 모든 인수 테스트를 통과시켜야 합니다.
- [x] 한 파일에는 하나의 컴포넌트만 있어야 합니다.
- [x] 테스트 커버리지 100%를 달성해야 합니다.

smilee committed:
Add test code

Add test code for `App`.

smilee committed:
Add store

Add store to manage state.

smilee committed:
Add test code

Add test code to test whether `RegionsContainer` renders.

smilee committed:
Add code

Add code to render `RegionsContainer`.

smilee committed:
Add test code

Add test code to see whether `CategoriesContainer` renders.

smilee committed:
Add code

Add code to render `CategoriesContainer`.

smilee committed:
Add test code

Add test code to see whether `RestaurantsContainer` renders.

smilee committed:
Add code

Add code to render `RestaurantsContainer`.

smilee committed:
Add test code

Add test code to see whether `ButtonContainer` renders.

smilee committed:
Add code

Add code to render `ButtonContainer`.

smilee committed:
Add test code

Add test code to see whether `Regions` renders.

smilee committed:
Add code

Add code to render `Regions`.

smilee committed:
Add test code

Add test code to see whether `Categories` renders.

smilee committed:
Add code

Add code to render `Categories`.

smilee committed:
Add test code

Add test code to see whether `Button` renders.

smilee committed:
Add code

Add code to render `Button`.

hannut91 reviewed (commented):
굉장히 단순한 케이스부터 시작해서 한 단계씩 진행하고 계신 게 인상적이네요 ㅎㅎ😀

smilee committed:
Add test code

Add test code to see whether `Button` handles click.

smilee committed:
Add code

Add code for `Button` to handle click.

smilee committed:
Add test code

Add test code for `reducer`.

smilee committed:
Add code

Add code for `reducer`.

smilee committed:
Add test code

Add test code for context, with action.

smilee committed:
Add code

Add code for context, with action.

smilee committed:
Add test code

Add test code for dispatch.

smilee committed:
Add code

Add code to load restaurants.

smilee committed:
Add test code

Add test code to see whether restaurants render.

smilee committed:
Add code

Add code to pass tests.

smilee committed:
Refactor `RestaurantContainer`

Refactor `RestaurantContainer` to render `Restaurants`.

smilee committed:
Add test code

Add test code for `setRestaurants`.

smilee committed:
Add code

Add code for `setRestaurants`.

smilee committed:
Add api

Add api to fetch data.

smilee committed:
Add filter

Add filter to dynamically show relevant restaurants.

smilee commented:
이제 Redux Thunk를 적용할 차례입니다.

hannut91 commented on src/Button.jsx:
```diff
@@ -0,0 +1,13 @@
+      onClick={onClick}
+    >
+      {name}
+      {selected ? '(V)' : null}
```
이렇게 써줘도 될거에요

```js
{selected && '(V)'}
```

smilee replied:
적용했습니다.

hannut91 commented on src/Button.test.jsx:
```diff
@@ -0,0 +1,58 @@
+    };
+  }
+
+  context('while not selected', () => {
```
context에서는 `with, without, when`만 쓸 것은 권장합니다.

See also
  - https://www.betterspecs.org/#contexts

smilee replied:
저 리스트에 `while`을 추가해도 괜찮을 것 같지만 그냥 `when`으로 바꿨습니다.

hannut91 commented on src/Button.test.jsx:
```diff
@@ -0,0 +1,58 @@
+  });
+
+  const name = '서울';
+  const onClick = jest.fn();
```
버튼이 클릭 됐을 때, 그 이벤트를 다루는 함수의 이름이니 `handleClick`이 맞을 것 같아요

smilee replied:
지금 컨테이너 컴포넌트에서 `handleClick`을 `onClick`으로 받고있는데, `handleClick`으로 받는 것이 더 좋을까요?

hannut91 replied:
컨테이에서는 onClick이란 이름으로 props를 받는게 맞고, 그 처리를 하는 함수 이름이 handleClick 입니다.

smilee replied:
다시 보고 이해했습니다!

hannut91 commented on src/reducer.test.jsx:
```diff
@@ -0,0 +1,97 @@
+          }),
+        );
+
+        expect(state.regions).toHaveLength(7);
```
왜 length가 7인지 이해하기 어려운 것 같아요. length 비교를 `regions.length`로 비교하거나 혹은

```js
expect(state.regions).toEqual(regions);
```

라고 써줘도 될 것 같아요

hannut91 commented on src/reducer.test.jsx:
```diff
@@ -0,0 +1,97 @@
+          }),
+        );
+
+        expect(state.categories).toHaveLength(7);
```
여기도 마찬가지로 왜 `7`인지 이해하기 어려운 것 같아요

hannut91 commented on src/reducer.test.jsx:
```diff
@@ -0,0 +1,97 @@
+          }),
+        );
+
+        expect(state.restaurants).toHaveLength(3);
```
여기도 마찬가지로 왜 `3`인지 이해하기 어려운 것 같아요

hannut91 commented on src/RestaurantsContainer.test.jsx:
```diff
@@ -0,0 +1,25 @@
+      restaurants,
+    }));
+  });
+  it('renders', () => {
```
여기 위에 빈 줄을 추가해서 구분해주면 좋을 것 같아요

hannut91 commented on src/Restaurants.jsx:
```diff
@@ -0,0 +1,13 @@
+export default function Restaurants({ restaurants }) {
+  return (
+    <ul>
+      {
+        restaurants && restaurants.map(({ id, name }) => (
+          <li key={id}>{name}</li>
+        ))
+      }
```
```js
restaurants?.map(({ id, name }) => (
  <li key={id}>{name}</li>
))
```

이렇게 써줘도 올바르게 동작할까요? 아마 undefined가 나와서 어떻게 화면에 보여질지는 잘 모르겠네요

See also
  - https://ko.javascript.info/optional-chaining

smilee replied:
동작하네요! 전에는 따로 설정을 해줘야 했던 것 같은데.. ㅎㅎ

hannut91 commented on src/ButtonContainer.jsx:
```diff
@@ -0,0 +1,25 @@
+  return (
+    <Button
+      name={name}
+      selected={filter && (name === filter.regionName || id === filter.categoryId)}
```
버튼이 두 가지 일을 해서 이렇게 된 것 같아요. regions container와 categories container를 나누는건 어떨까요?

smilee replied:
그렇게 하는 것도 좋은 것 같습니다. 특히 처리하는 종류가 더 다양해지면 나누는 것이 더 좋을 것 같습니다. 😄

smilee committed:
Refactor code

Refactor code to improve code comprehension.

smilee committed:
Implement redux-thunk

Implement redux-thunk to write action creators that return a function instead of an action.

smilee committed:
Refactor code

Refactor code to cover more cases.
